### PR TITLE
qgis: 3.10.1 -> 3.10.4 (+ libs)

### LINKED ISSUE #78755: upgrade QGIS to 3.10.2
Both, [nixos-19.09](https://nixos.org/nixos/packages.html?attr=qgis&channel=nixos-19.09&query=qgis) and [nixos-unstable currently ship QGIS 3.10.**1**](https://nixos.org/nixos/packages.html?attr=qgis&channel=nixos-unstable&query=qgis).

In their blog Post ["Public Service Announcement: Update to the latest point release now"](http://blog.qgis.org/2020/01/24/public-service-announcement-update-to-the-latest-point-release-now/), the QGIS core developers recommend all QGIS 3.10.x users to update to QGIS 3.10.**2** as soon as possible:

> QGIS users who have adopted the 3.10 version when initially released at the end of October 2019 have likely noticed a sharp drop in reliability. The underlying issues have now been addressed in 3.10.2, all users are advised to update *now*.

Some of the fixes were in QGIS itself, some in the underlying libraries GDAL/OGR and PROJ. So these should be upgraded as well:

> When QGIS 3.10 was first released in the end of October 2019, a pair of libraries – namely GDAL and PROJ – were updated to their next-generation versions. [...]
>
> The update of these crucial libraries led to a number of regressions. While we expected some issues to arise, the seriousness of the disruption caught us off guard. Yet, it was also somewhat inevitable: QGIS is the first large GIS project to expose these next-generation libraries to the masses. The large number of QGIS users across the globe were essentially stress testing both new code within QGIS as well as the libraries themselves.

("next-generation versions" probably refers to a change in the major versions.)

The versions that should be used are

* GDAL/OGR 3.0.3
* PROJ 6.3.0

as I've figured out in [my [QGIS-Developer] mailing list post here](https://lists.osgeo.org/pipermail/qgis-developer/2020-January/060056.html).

Both nixos-19.09 and nixos-unstable currently ship PROJ 6.1.1 and GDAL/OGR 3.0.1.

Thus, I see the following things to do:

- [x] `gdal`: 3.0.1 -> 3.0.3 on `master` (PR #78774)
- [ ] `proj`: 6.1.1 -> 6.3.0 on `master` (PR #83317)
- [ ] `qgis`: 3.10.1 -> 3.10.2 on `master` (PR #83317)

- [x] `gdal`: 3.0.1 -> 3.0.3 on `release-19.09` (PR #79257)
- [ ] `proj`: 6.1.1 -> 6.3.0 on `release-19.09`
- [ ] `qgis`: 3.10.1 -> 3.10.2 on `release-19.09`

/cc qgis maintainer @lsix
/cc gdal maintainer @MarcWeber
/cc proj maintainer @vbgl

### PR DESCRIPTION
**Pull request by lsix**
###### Motivation for this change

Update QGIS.

In this PR, I do not upgrade `proj` up to 7.0.0 since I intend to backport those changes to `release-20.03`. I’ll do the `proj-7.0.0` update after in a separate PR.

The update of `pyproj` makes it python2 incompatible and few packages become unavailable in python2Packages as a consequence, but I do not think this is a problem (python2 is deprecated anyway).

This closes #78755 

cc @das-g 

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pyproj/001.proj.patch [hunk 1]
```diff
@@ -1,47 +1,64 @@
...
+ 
+@@ -17,6 +18,16 @@
+     with tempfile.TemporaryDirectory() as tmpdir:
+         tmp_data_dir = os.path.join(tmpdir, "proj")
+         shutil.copytree(data_dir, tmp_data_dir)
++        # Data copyed from the store are readonly (causes cleanup problem).
++        # Make they are writable
++        for r, d, files in os.walk(tmp_data_dir):
++            stat_info = os.stat(r)
++            os.chmod(r, stat_info.st_mode | stat.S_IWUSR)
```
Comment (das-g) [inline pkgs/development/python-modules/pyproj/001.proj.patch]: Variable `stat_info` could be inlined without much loss of comprehensibility.
```suggestion
+            os.chmod(r, os.stat(r).st_mode | stat.S_IWUSR)
```

File: pkgs/development/python-modules/pyproj/001.proj.patch [hunk 2]
```diff
@@ -1,47 +1,64 @@
...
++            for f in files:
++                fpath = os.path.join(r, f)
++                stat_info = os.stat(fpath)
++                os.chmod(fpath, stat_info.st_mode | stat.S_IWUSR)
```
Comment (das-g) [inline pkgs/development/python-modules/pyproj/001.proj.patch]: Variable stat_info could be inlined without much loss of comprehensibility.
```suggestion
+                os.chmod(fpath, os.stat(fpath).st_mode | stat.S_IWUSR)
```

**Review by das-g — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/pyproj/001.proj.patch [hunk 3]
```diff
@@ -1,47 +1,64 @@
...
++
+         try:
+             os.remove(os.path.join(str(tmp_data_dir), "ntv2_0.gsb"))
+         except OSError:
+Seulement dans b/test: .conftest.py.swp
```
[the remark this answers is not in the document]
Comment (lsix) [reply pkgs/development/python-modules/pyproj/001.proj.patch]: Oups, I did not mean to include those :sweat_smile:

File: pkgs/development/python-modules/pyproj/default.nix [hunk 4]
```diff
@@ -38,6 +41,9 @@ buildPythonPackage rec {
   checkPhase = ''
     pytest . -k 'not alternative_grid_name \
                  and not transform_wgs84_to_alaska \
+                 and not transformer_group__unavailable \
+                 and not transform_group__missing_best \
+                 and not datum \
```
[the remark this answers is not in the document]
Comment (lsix) [reply pkgs/development/python-modules/pyproj/default.nix]: I did not manage to have those working (unfortunatly). They are related to rounding errors when using the `NAD27` projection.

**Review by das-g — COMMENTED** [review]
:heavy_check_mark: Diff LGTM. :+1:

:thinking: I wonder whether `proj: 6.1.1 -> 6.3.1` (currently cbb385a) and `pythonPackages.pyproj: 2.2.2 -> 2.6.0` (currently da164cb) should instead be done in a **single** commit, as one AFAIK cannot bump one without breaking the other.

:hourglass_flowing_sand: I didn't test it, yet.

**Comment by lsix** [comment]
Should be better.

**Comment by lsix** [comment]
I think (but not tested yet, I’ll have a look when I’ll prepare the `release-20.03` backport) that I can only disable the tests with rounding issues in pyproj (arising in both versions) and it should work. But I’ll check when preparing the backport since I would prefer not to update pyproj since it removes python2 packages.

**Review by jonringer — APPROVED** [review]
diff LGTM
commits LGTM

```
[68 built, 300 copied (2511.5 MiB), 497.3 MiB DL]
https://github.com/NixOS/nixpkgs/pull/83317
8 package marked as broken and skipped:
gazebo gazebo-headless gazeboSimulator.gazebo6 gazeboSimulator.gazebo6-headless gazeboSimulator.gazebo7 gazeboSimulator.gazebo7-headless gplates openspace

56 package built:
gdal gdal_1_11 gdal_2 grass libLAS libgeotiff libspatialite mapnik merkaartor mysql-workbench openorienteering-mapper osm2pgsql osm2xmap perl528Packages.Tirex perl530Packages.Tirex postgresqlPackages.postgis proj python27Packages.fiona python27Packages.gdal python27Packages.labelbox python27Packages.python-mapnik python27Packages.rasterio python27Packages.tilestache python27Packages.worldengine python37Packages.basemap python37Packages.cartopy python37Packages.fiona python37Packages.geopandas python37Packages.labelbox python37Packages.osmnx python37Packages.owslib python37Packages.pyproj python37Packages.python-mapnik python37Packages.rasterio worldengine-cli python38Packages.basemap python38Packages.cartopy python38Packages.fiona python38Packages.gdal python38Packages.geopandas python38Packages.labelbox python38Packages.osmnx python38Packages.owslib python38Packages.pyproj python38Packages.python-mapnik python38Packages.rasterio python38Packages.worldengine qgis qgis-unwrapped qlandkartegt qmapshack readosm saga spatialite_tools xastir xygrib
```

**Comment by jonringer** [comment]
> thinking I wonder whether `proj: 6.1.1 -> 6.3.1` (currently [cbb385a](https://github.com/NixOS/nixpkgs/commit/cbb385a052cdf58f3f78707da4b4241f2bec3753)) and `pythonPackages.pyproj: 2.2.2 -> 2.6.0` (currently [da164cb](https://github.com/NixOS/nixpkgs/commit/da164cbba59882a2deb41cbf278b9b88ded55f9e)) should instead be done in a **single** commit, as one AFAIK cannot bump one without breaking the other.

I would have them as separate commits, as it might help to bisect if the bump in pyproj or proj caused a regression

**Pull request merged**